testcases/kernel/device-drivers/common/common.sh

```diff
@@ -0,0 +1,425 @@
+  v1=`expr $1 + 1`
+  expr $v % $v1
+}
+
+# random not equal to 0
+random_ne0()
+{
+  v=`_random $1`
+  #v=`dd if=/dev/urandom count=1 2> /dev/null | cksum | cut -c 0-10`
+  expr $v % $1 + 1
+}
+
+# check different kernel errors
+check_kernel_errors()
+{
+  local type=$1
+  shift
+  local opts=$1
+  case $type in
+    kmemleak)
+      check_config_options "y" CONFIG_DEBUG_KMEMLEAK
+      kmemleaks="/sys/kernel/debug/kmemleak"
+      if [ ! -e ${kmemleaks} ]; then
+        die "kmemleak sys entry doesn't exist; check if CONFIG_DEBUG_KMEMLEAK=y in kernel configs; perhaps need to increase CONFIG_DEBUG_KMEMLEAK_EARLY_LOG_SIZE"
+      fi
+
+      if [ "$opts" = "clear" ]; then
+        # clear the list of all current possible memory leaks before scan
+        do_cmd "echo clear > ${kmemleaks}"
+        return
+      fi
+
+      # trigger memory scan
+      do_cmd "echo scan > ${kmemleaks}"
+      # give kernel some time to scan
+      do_cmd sleep 30
+      kmemleak_detail=`cat ${kmemleaks}`
+      if [ -n "${kmemleak_detail}" ]; then
+        die "There are memory leaks being detected. The details are displayed as below: ${kmemleak_detail}"
+      else
+        test_print_trc "No memory leaks being detected."
+      fi
+
+    ;;
+    spinlock)
+      check_config_options "y" CONFIG_DEBUG_SPINLOCK
+
+      if [ "$opts" = "clear" ]; then
+        dmesg -c
+        return
+      fi
+
+      # Check dmesg to catch the error
+      spinlock_errors="BUG: spinlock"
+      dmesg |grep -i "${spinlock_errors}" && die "There is spinlock errors showing in dmesg" || test_print_trc "No spinlock related error found in dmesg"
+    ;;
+    *)
+      die "check_kernel_errors: No logic for type $type yet."
+    ;;
+  esac
+}
+
+# Function to check if environment variable is set
+# Input: $1 - environment variable to be checked
+# returns true if set
+# returns with error if not set
+check_env_var() {
+  output_str=`env|grep $1`
+  if [ ${#output_str} == 0 ]
+  then
+    die "$1 not defined"
+  fi
+}
+
+# $1: check type, either 'y', 'm', 'ym' or 'n'
+# $2: Options to check. Uses same syntax returned by get_modular_config.names.sh
+#     which is CONFIG1^CONFIG2:module1 CONFIG3:module2
+check_config_options()
+{
+  config_cmd='zcat /proc/config.gz'
+  ls /boot/config-`uname -r` &> /dev/null && config_cmd="cat /boot/config-"`uname -r`
+  case $1 in
+    y) check='=y';;
+    m) check='=m';;
+    ym) check='(=y|=m)';;
+    n) check=' is not set';;
+    *) die "$1 is not a valid check_config_options() option"
+  esac
+  OIFS=$IFS
+  IFS=' '
+  shift
+  x=$*
+  x=${x[@]}
+  y=()
+  for i in $x
+  do
+    newval=`echo $i | cut -d':' -f 1`
+    y+=($newval)
+  done
+  IFS='^';y=${y[@]}
+  IFS=' '
+  for option in $y; do
+     $config_cmd | egrep "$option$check" && continue ||  (IFS=$OIFS; \
+     if is_opt_config "$option"; then skip_test "optional $option not enabled"; else die "$option is not $check"; fi); exit $?
+  done
+  IFS=$OIFS
+}
```

Comment:
LTP has meanwhile implemented an API for describing kernel config requirements, make use of that instead.